src/useRouteQueryParam/formats/RouteParam.ts

```diff
@@ -0,0 +1,105 @@
+import { LocationQueryValue } from 'vue-router'
+import { UseRouteQuery } from '@/useRouteQuery/useRouteQuery'
+import { InvalidRouteParamValue, isInvalidRouteParamValue, isNotInvalidRouteParamValue } from '@/useRouteQueryParam/formats/InvalidRouteParamValue'
+import { asArray } from '@/utilities/arrays'
+
+const IS_ROUTE_PARAM_SYMBOL: unique symbol = Symbol()
+
+export type RouteParamClassArgs<T> = {
+  key: string,
+  defaultValue: T | T[],
+}
+
+// adding any here so RouteParamClass can be used without passing a generic
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type RouteParamClass<T = any> = new (args: RouteParamClassArgs<T>) => RouteParam<T>
+
+export function isRouteParamClass(value: unknown): value is RouteParamClass<unknown> {
+  return typeof value === 'function' && 'IS_ROUTE_PARAM' in value && value.IS_ROUTE_PARAM == IS_ROUTE_PARAM_SYMBOL
+}
+
+export abstract class RouteParam<T> {
+  public static IS_ROUTE_PARAM = IS_ROUTE_PARAM_SYMBOL
+
+  protected abstract parse(value: LocationQueryValue): T
+  protected abstract format(value: T): LocationQueryValue
+
+  protected key: string
+  protected defaultValue: T | T[]
+
+  private get multiple(): boolean {
+    return Array.isArray(this.defaultValue)
+  }
+
+  public constructor({ key, defaultValue }: RouteParamClassArgs<T>) {
+    this.key = key
+    this.defaultValue = defaultValue
+  }
+
+  public get(routeQuery: UseRouteQuery): T | T[] {
+    if (!(this.key in routeQuery.query)) {
+      return this.defaultValue
+    }
+
+    const strings = asArray(routeQuery.get(this.key))
+    const values = strings.map(value => this.safeParseValue(value)).filter(isNotInvalidRouteParamValue)
+
+    if (this.multiple) {
+      return values
+    }
+
+    const [first] = values
+
+    return first
+  }
+
+  public set(routeQuery: UseRouteQuery, value: T | T[] | undefined): void {
+    if (value === undefined) {
+      return routeQuery.remove(this.key)
+    }
+
+    const values = asArray(value)
+    const strings = values.map(value => this.safeFormatValue(value)).filter(isNotInvalidRouteParamValue)
+
+    if (this.multiple) {
+      this.defaultValue = []
+    }
```

Comment:
why update the defaultValue here?

Reply:
I have no idea.... I feel like this was left over from a previous iteration. I'll look into it but I think this should just be removed.

Reply:
Ohh I remember. This was from a previous iteration but the thought is still valid. The thought being, if there is a default value for an array param. That default is used if the param does not exist. But that also means there's no way to get to an empty array. 

I haven't rethought about that problem since the last refactor. So taking a chance to think about it again. What are your thoughts on that?

Reply:
if that is in case needed, we should squirrel this away in a function that explains what we're up to

Reply:
This needs some more work. But for now I'm going to remove these lines and we can support the use case like this
```typescript
const states = useRouteQueryParam('state', [])

if(states.length === 0) {
  states.value = ['running'] // this is the "default"
}
```